### PR TITLE
Add support for custom authorization request parameters

### PR DESCRIPTION
**Pull request by jgrandja**
This PR adds support for appending additional request parameters on top of the standard [OAuth 2.0 Authorization Request parameters](https://tools.ietf.org/html/rfc6749#section-4.1.1).

OpenID Connect 1.0 defines [Authentication Request parameters](http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) on top of the OAuth 2.0 Authorization Request parameters that this feature will be able to support.

Providers may also define non-standard parameters which this feature will be able to support as well.

This PR fixes gh-4911

### CODE REVIEW

**Comment by rwinch** [comment]
@jgrandja For record keeping purposes. As discussed offline, we can resolve this by placing query parameters in the authorization URL. I'm marking this as waiting-for-changes as I don't think we can safely close it and that is the closest status we have for pending closure

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2AuthorizationRequestRedirectFilter.java [hunk 1]
```diff
@@ -112,14 +100,23 @@ public OAuth2AuthorizationRequestRedirectFilter(ClientRegistrationRepository cli
...
+				clientRegistrationRepository, authorizationRequestBaseUri);
+	}
+
+	/**
+	 * Sets the resolver used for resolving {@link OAuth2AuthorizationRequest}'s.
+	 *
+	 * @since 5.1
+	 * @param authorizationRequestResolver the resolver used for resolving authorization requests
+	 */
+	public final void setAuthorizationRequestResolver(OAuth2AuthorizationRequestResolver authorizationRequestResolver) {
```
Comment (rwinch) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2AuthorizationRequestRedirectFilter.java]: This should be a separate constructor. The reason is that with a custom OAuth2AuthorizationRequestResolver implementation, the ClientRegistrationRepository should not be required for this class. However, it is required to construct the class at the moment.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java [hunk 2]
```diff
@@ -50,6 +56,7 @@
 	private Set<String> scopes;
 	private String state;
 	private Map<String, Object> additionalParameters;
+	private URI authorizationRequestUri;
```
Comment (rwinch) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java]: Change to String for consistency.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java [hunk 3]
```diff
@@ -144,6 +161,64 @@ public static Builder implicit() {
...
+	}
+
+	/**
+	 * Returns a {@code URI} representation for the provided {@code authorizationRequest}.
+	 *
+	 * @since 5.1
+	 * @param authorizationRequest the authorization request
+	 * @return a {@code URI} representation of the authorization request
+	 */
+	public static URI asUri(OAuth2AuthorizationRequest authorizationRequest) {
```
Comment (rwinch) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java]: This method does not make sense to me. What is it constructing? The OAuth2AuthorizationRequest has a authorizationRequestUri property, so it is illogical to pass in an argument that contains the value that you are producing. I think this logic belongs encapsulated within the resolver and we can open it up later if necessary.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2AuthorizationRequestRedirectFilter.java (hunk 1 above)
Comment (jgrandja) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2AuthorizationRequestRedirectFilter.java]: Good catch. Makes sense.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java (hunk 2 above)
Comment (jgrandja) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java]: If we store this as a `String` than is it the encoded or decoded version? The `OAuth2AuthorizationRequestRedirectFilter` and `OAuth2AuthorizationRequestRedirectWebFilter` will need to know this before it initiates the redirect.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java (hunk 3 above)
Comment (jgrandja) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/OAuth2AuthorizationRequest.java]: > The OAuth2AuthorizationRequest has a authorizationRequestUri property, so it is illogical to pass in an argument that contains the value that you are producing

Yeah good point

> I think this logic belongs encapsulated within the resolver

Why not keep this logic in `OAuth2AuthorizationRequest` encapsulated. It provides the default `authorizationRequestUri` in `build()` if the user has not supplied it.

```
			authorizationRequest.authorizationRequestUri =
					this.authorizationRequestUri != null ?
						this.authorizationRequestUri : asUri(authorizationRequest);

```

**Comment by jgrandja** [comment]
@rwinch I've applied all the updates from your recent feedback. However, I decided to leave the logic for building the `authorizationRequestUri` in `OAuth2AuthorizationRequest` (encapsulated) instead of the `DefaultOAuth2AuthorizationRequestResolver`. See me previous comment for my reasoning.

Also, it almost slipped by me as we need to add the ability to configure a custom `OAuth2AuthorizationRequestResolver` in `OAuth2LoginConfigurer` and `OAuth2ClientConfigurer`. Would you prefer a new ticket logged or bring it into this PR?

**Comment by rwinch** [comment]
This is now merged in via 779597af2a6ed777707f08ae8106818e0b8e299e

**Pull request closed without merging**